### ISSUE TITLE
Range of lines for fixers

### ISSUE THREAD

**Issue by IngoHeimbach** [w0rp/ale#850]
Hey, thanks for this great linter plugin. I switched from syntastic recently and it's amazing how fast and responsive real time linting works. :) 

In (this commit)[https://github.com/IngoHeimbach/ale/commit/fd92ca17e87eb3ec59fe57d42e52d1644264dd5e] I have added clang-tidy as a possible fixer for C and C++ code. However, often I only want to fix one specific line in a huge file (for example a missing semicolon or a missing `override`). Would it be possible to modify `ALEFix` to accept a specific line or a range of lines? To stay compatible with existing code fixers could get an optional third parameter with a range of lines they can operate on.

**Comment by w0rp** [comment]
I wonder if this use case is worth supporting. Not all tools support fixing ranges of lines. You often have to look at an entire file to know if you can fix something. A program for fixing errors should only really make changes where there are problems. Otherwise, why change anything?

**Comment by bb010g** [comment]
[google/google-java-format](https://github.com/google/google-java-format) supports a similar argument.

**Comment by w0rp** [comment]
I recommend writing smaller files.

**Comment by WoLpH** [comment]
I generally don't need the fixers for code I wrote ;) And I can't control what my customers have running in all cases

**Comment by sherwinyu** [comment]
I agree that this would be a useful feature. I was excited about using `:ALEFix` and a python fixer (autopep8 / yapf) on my codebase at work, but ended up disabling the fix-on-write behavior because it introduce a lot of unrelated style fixes, since it's changing the entire file, not just the lines I touched. Frequently I want to limit it to just the changes I've made. 

Supporting this would be incredible!

My current workaround is something like this:  `:'<,'>!autopep8 -` to format the selection

**Comment by w0rp** [comment]
Someone needs to figure out how to fit this into the existing API, including how fixer functions should be defined. Considerations need to be made when combining tools which do support fixing ranges with those that do not.

**Comment by WoLpH** [comment]
Agreed, especially with languages such as python which are largely relying on whitespace this can be dangerous

**Comment by renstrom** [comment]
Some tools like `clang-format` and `yapf` (Python) takes a "range" argument, but actually require the whole file for context.

>My current workaround is something like this: `:'<,'>!autopep8 -` to format the selection

I'm curious, is this even working as you expect given what I wrote in my first paragraph? As far as I know you need to pass the `--line-range`/`--range` to `autopep8`:

```
  --line-range line line, --range line line
                        only fix errors found within this inclusive range of
                        line numbers (e.g. 1 99); line numbers are indexed at
                        1
```

**Comment by WoLpH** [comment]
That works, assuming you select an entire function/class or other top-level code. Which already helps a lot imho :)

It beats having to format everything yourself so I use similar tricks

**Comment by w0rp** [comment]
I've thought about how to support this a bit, and here is my plan.

1. Add commands for fixing ranges of lines which work with selection modes.
2. When a range command is used, send `[buffer_number, [lnum, col], [end_lnum, end_col]]` to the functions as the first fixer function argument instead of just `buffer_number`.
3. Expect functions to return a `Dictionary` containing the key `fix_range` set to `1` for fixing a range. For applying a list immediately, look for the `new_lines` key. Ignore all other output.
4. Consider catching E745 only when fixing ranges of lines, so problems with handling the List for other functions can be ignored.

With the above it would be possible to fix ranges of lines for the functions that support it, and ignore the output of others. You should expect weird results when mixing fixers which do and do not support fixing ranges.

**Comment by leikoilja** [comment]
@w0rp, is there any update for us? It would be so so cool to be able to send a range of lines for a python linter.
Thanks

**Comment by plasmon360** [comment]
> I agree that this would be a useful feature. I was excited about using `:ALEFix` and a python fixer (autopep8 / yapf) on my codebase at work, but ended up disabling the fix-on-write behavior because it introduce a lot of unrelated style fixes, since it's changing the entire file, not just the lines I touched. Frequently I want to limit it to just the changes I've made.
> 
> Supporting this would be incredible! (And thanks for the already incredible plugin)
> 
> My current workaround is something like this: `:'<,'>!autopep8 -` to format the selection

Thanks. I am using :'<,'>!black-macchiato . [black-macchiato works on top of black](https://github.com/wbolster/black-macchiato)
